File: backend/app/api/routes/collector_map.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.dependencies import get_db, require_roles
from app.models.user import User
from app.schemas.collector import (
    CollectorLocationRead,
    CollectorLocationUpdate,
    CollectorMapRead,
    NavigationRead,
    RouteSummaryRead,
)
from app.schemas.pickup_request import NearbyPickupRequestRead
from app.services.collector_map import (
    get_collector_location,
    get_collector_map,
    get_collector_route,
    get_navigation_route,
    list_nearby_pickups,
    update_collector_location,
)

router = APIRouter()


@router.get("/map", response_model=CollectorMapRead)
def collector_map(
    latitude: float | None = Query(None, ge=-90, le=90),
    longitude: float | None = Query(None, ge=-180, le=180),
    radius_km: float = Query(5, gt=0, le=200),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles("collector")),
) -> CollectorMapRead:
    return get_collector_map(db, current_user, latitude, longitude, radius_km)


@router.get("/location", response_model=CollectorLocationRead)
def collector_location(
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles("collector")),
) -> CollectorLocationRead:
    return get_collector_location(db, current_user)


@router.post("/location", response_model=CollectorLocationRead)
def report_collector_location(
    payload: CollectorLocationUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles("collector")),
) -> CollectorLocationRead:
    return update_collector_location(db, current_user, payload)


@router.get("/route", response_model=RouteSummaryRead)
def collector_route(
    latitude: float | None = Query(None, ge=-90, le=90),
    longitude: float | None = Query(None, ge=-180, le=180),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles("collector")),
) -> RouteSummaryRead:
    return get_collector_route(db, current_user, latitude, longitude)


@router.get("/nearby-pickups", response_model=list[NearbyPickupRequestRead])
def nearby_pickups(
    latitude: float | None = Query(None, ge=-90, le=90),
    longitude: float | None = Query(None, ge=-180, le=180),
    radius_km: float = Query(5, gt=0, le=200),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles("collector")),
) -> list[NearbyPickupRequestRead]:
    return list_nearby_pickups(db, current_user, latitude, longitude, radius_km)


@router.get("/navigation/{pickup_id}", response_model=NavigationRead)
def collector_navigation(
    pickup_id: int,
    latitude: float | None = Query(None, ge=-90, le=90),
    longitude: float | None = Query(None, ge=-180, le=180),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles("collector")),
) -> NavigationRead:
    return get_navigation_route(db, current_user, pickup_id, latitude, longitude)

File: frontend/src/api/collectorMap.ts
import apiClient from "@/api/client";
import type {
  CollectorLocation,
  CollectorLocationUpdatePayload,
  CollectorMapPayload,
  Navigation,
  NearbyPickup,
  RouteSummary,
} from "@/types/map";

export interface CollectorMapParams {
  latitude?: number;
  longitude?: number;
  radiusKm?: number;
}

export async function getCollectorMap(
  params: CollectorMapParams = {},
): Promise<CollectorMapPayload> {
  const response = await apiClient.get<CollectorMapPayload>("/collector/map", {
    params: {
      latitude: params.latitude,
      longitude: params.longitude,
      radius_km: params.radiusKm,
    },
  });
  return response.data;
}

export async function getCollectorLocation(): Promise<CollectorLocation> {
  const response = await apiClient.get<CollectorLocation>("/collector/location");
  return response.data;
}

export async function updateCollectorLocation(
  payload: CollectorLocationUpdatePayload,
): Promise<CollectorLocation> {
  const response = await apiClient.post<CollectorLocation>("/collector/location", payload);
  return response.data;
}

export async function getCollectorRoute(
  params: { latitude?: number; longitude?: number } = {},
): Promise<RouteSummary> {
  const response = await apiClient.get<RouteSummary>("/collector/route", {
    params: {
      latitude: params.latitude,
      longitude: params.longitude,
    },
  });
  return response.data;
}

export async function listNearbyPickups(
  params: { latitude?: number; longitude?: number; radiusKm?: number } = {},
): Promise<NearbyPickup[]> {
  const response = await apiClient.get<NearbyPickup[]>("/collector/nearby-pickups", {
    params: {
      latitude: params.latitude,
      longitude: params.longitude,
      radius_km: params.radiusKm,
    },
  });
  return response.data;
}

export async function getCollectorNavigation(
  pickupId: number,
  params: { latitude?: number; longitude?: number } = {},
): Promise<Navigation> {
  const response = await apiClient.get<Navigation>(`/collector/navigation/${pickupId}`, {
    params: {
      latitude: params.latitude,
      longitude: params.longitude,
    },
  });
  return response.data;
}